kvbc/include/db_interfaces.h

```diff
@@ -35,13 +42,85 @@ class ILocalKeyValueStorageReadOnly {
+
+  virtual ~IBlockAdder() = default;
+};
+
+// Key-value blockchain categorized read interface.
+//
+// Output vector parameters to multi* calls can contain more values than requested. This is an optimization, in order to
+// reduce memory allocations. Users are encouraged to reuse a single vector instance.
+class IReader {
+ public:
+  // Get the value of a key in `category_id` at `block_id`.
+  // Return std::nullopt if `key` doesn't exist at `block_id`.
+  virtual std::optional<categorization::Value> get(const std::string &category_id,
+                                                   const std::string &key,
+                                                   BlockId block_id) const = 0;
+
+  // Get the latest value of `key` in `category_id`.
+  // Return std::nullopt if the key doesn't exist or is deleted.
+  virtual std::optional<categorization::Value> getLatest(const std::string &category_id,
+                                                         const std::string &key) const = 0;
+
+  // Get values for keys at specific versions in `category_id`.
+  // `keys` and `versions` must be the same size.
+  // If a key is missing at the specified version or deleted, then std::nullopt is returned for it.
+  virtual void multiGet(const std::string &category_id,
+                        const std::vector<std::string> &keys,
+                        const std::vector<BlockId> &versions,
+                        std::vector<std::optional<Value>> &values) const = 0;
+
+  // Get the latest values of a list of keys in `category_id`.
+  // If a key is missing or is deleted, then std::nullopt is returned for it.
+  virtual void multiGetLatest(const std::string &category_id,
+                              const std::vector<std::string> &keys,
+                              std::vector<std::optional<Value>> &values) const = 0;
+
+  // Get the latest version of `key` in `category_id`.
+  // Return std::nullopt if the key doesn't exist or is deleted.
+  virtual std::optional<categorization::TaggedVersion> getLatestVersion(const std::string &category_id,
+                                                                        const std::string &key) const = 0;
+
+  // Get the latest versions of the given keys in `category_id`.
+  // If a key is missing, then std::nullopt is returned for its version.
+  virtual void multiGetLatestVersion(const std::string &category_id,
+                                     const std::vector<std::string> &keys,
+                                     std::vector<std::optional<categorization::TaggedVersion>> &versions) const;
+
+  // Get the value of `key` and a proof for it at `block_id`.
+  // Return std::nullopt if the key doesn't exist.
+  virtual std::optional<categorization::Proof> getProof(const std::string &category_id,
+                                                        const std::string &key,
+                                                        BlockId block_id) const = 0;
+
```

Comment:
Maybe this should be defined not as part of the read interface? maybe IProof ?

Reply:
I don't have a strong preference. My assumption was that users of read will also be the ones that want proofs.

Reply:
I'm in favor of small interfaces that are easy to implement, moreover it proof is not really a read operation, it actually calculates stuff.

Reply:
Actually, I am thinking of removing the proof part for now, because there are some differences (there are tags for the immutable category, possibly differences in the block merkle proof) and maybe this is not thought through yet. Will just leave the read part and handle proofs separately.

Reply:
Removed proofs altogether.